DirectProgramming/C++SYCL_FPGA/Tutorials/Features/task_sequences/task_sequence/src/main.cpp

```diff
@@ -0,0 +1,206 @@
+    kPipeMinCapacity  // The capacity of the pipe
+    >;
+
+class IDPipeOut;
+using PipeOut = sycl::ext::intel::experimental::pipe<
+    IDPipeOut,        // An identifier for the pipe
+    int,              // The type of data in the pipe
+    kPipeMinCapacity  // The capacity of the pipe
+    >;
+
+[[intel::use_stall_enable_clusters]] 
+void LoopA(int len) {
+  [[intel::initiation_interval(1)]]  
+  for (size_t i = 0; i < len; i++) {
+    int in0 = PipeIn0::read();
+    int in1 = PipeIn1::read();
+
+    PipeAB::write(in0);
+    PipeAD::write(in1);
+  }
+}
+
+[[intel::use_stall_enable_clusters]] 
+void LoopB(int len) {
+  [[intel::initiation_interval(1)]]  
+  for (size_t i = 0; i < len; i++) {
+    int tmp = PipeAB::read();
+    tmp += i;
+    PipeBC::write(tmp);  
+  }
+}
+
+[[intel::use_stall_enable_clusters]] 
+void LoopC(int len) {
+  [[intel::initiation_interval(1)]]  
+  for (size_t i = 0; i < len; i++) {
+    int tmp = PipeBC::read();
+    tmp += i;
+    PipeCD::write(tmp);  
+  }
+}
+
+[[intel::use_stall_enable_clusters]] 
+void LoopD(int len) {
+  [[intel::initiation_interval(1)]]  
+  for (size_t i = 0; i < len; i++) {
+    int tmp0 = PipeCD::read();
+    int tmp1 = PipeAD::read();
+    int out = tmp0 + tmp1;
+    PipeOut::write(out);  
+  }
+}
+
+///////////////////////////////////////
+
+struct OptimizedKernel {
+  int len;
+
+  void operator()() const {
+    sycl::ext::intel::experimental::task_sequence<LoopA> task_a;
+    sycl::ext::intel::experimental::task_sequence<LoopB> task_b;
+    sycl::ext::intel::experimental::task_sequence<LoopC> task_c;
+    sycl::ext::intel::experimental::task_sequence<LoopD> task_d;
+
+    task_a.async(len);
+    task_b.async(len);
+    task_c.async(len);
+    task_d.async(len);
+  }
+};
+
+int main() {
+  bool passed = false;
+
+  try {
+    // Use compile-time macros to select either:
+    //  - the FPGA emulator device (CPU emulation of the FPGA)
+    //  - the FPGA device (a real FPGA)
+    //  - the simulator device
+#if FPGA_SIMULATOR
+    auto selector = sycl::ext::intel::fpga_simulator_selector_v;
+#elif FPGA_HARDWARE
+    auto selector = sycl::ext::intel::fpga_selector_v;
+#else  // #if FPGA_EMULATOR
+    auto selector = sycl::ext::intel::fpga_emulator_selector_v;
+#endif
+
+    sycl::queue q(selector, fpga_tools::exception_handler,
+                  sycl::property::queue::enable_profiling{});
+
+    auto device = q.get_device();
+
+    std::cout << "Running on device: "
+              << device.get_info<sycl::info::device::name>().c_str()
+              << std::endl;
+
+    int *a = new int[kVectSize];
+    int *b = new int[kVectSize];
+
+    // Generate input data
+    for (int i = 0; i < kVectSize; i++) {
+      a[i] = i;                
+      b[i] = (kVectSize - i);  
+
+      PipeIn0::write(q, i);
+      PipeIn1::write(q, kVectSize - i);
+    }
+
+    // Call the kernel
+    auto e = q.single_task<IDOptimized>(OptimizedKernel{kVectSize});
+    e.wait();
```

Comment:
any reason for putting wait for kernel to exit here? I see the naive code has the wait to happen after testbench. is this intended for both to be different placement?